lib/internal/main/run.js

```diff
@@ -0,0 +1,71 @@
+'use strict';
+/* eslint-disable node-core/prefer-primordials */
+
+// There is no need to add primordials to this file.
+// `run.js` is a script only executed when `node run <script>` is called.
+const {
+  prepareMainThreadExecution,
+  markBootstrapComplete,
+} = require('internal/process/pre_execution');
+const { getPackageJSONScripts } = internalBinding('modules');
+const { execSync } = require('child_process');
+const { resolve, delimiter } = require('path');
+
+prepareMainThreadExecution(false, false);
+
+markBootstrapComplete();
+
+// TODO(@anonrig): Search for all package.json's until root folder.
+const json_string = getPackageJSONScripts();
+
+// Check if package.json exists and is parseable
+if (json_string === undefined) {
+  process.exit(1);
+  return;
+}
+const scripts = JSON.parse(json_string);
+// Remove the first two arguments, which are the node binary and the command "run"
+const args = process.argv.slice(2);
+const id = args.shift();
+let command = scripts[id];
+
+if (!command) {
+  const { error } = require('internal/console/global');
+
+  error(`Missing script: "${id}"`);
+
+  const keys = Object.keys(scripts);
+  if (keys.length === 0) {
+    error('There are no scripts available in package.json');
+  } else {
+    error('Available scripts are:\n');
+    for (const script of keys) {
+      error(`  ${script}: ${scripts[script]}`);
+    }
+  }
+  process.exit(1);
+  return;
+}
+
+const env = process.env;
+const cwd = process.cwd();
+const binPath = resolve(cwd, 'node_modules/.bin');
+
+// Filter all environment variables that contain the word "path"
+const keys = Object.keys(env).filter((key) => /^path$/i.test(key));
+const PATH = keys.map((key) => env[key]);
+
+// Append only the current folder bin path to the PATH variable.
+// TODO(@anonrig): Add support for appending the bin path of all parent folders.
+const paths = [binPath, PATH].join(delimiter);
+for (const key of keys) {
+  env[key] = paths;
+}
+
+// If there are any remaining arguments left, append them to the command.
+// This is useful if you want to pass arguments to the script, such as
+// `node run linter --help` which runs `biome --check . --help`
+if (args.length > 0) {
+  command += ' ' + args.map((arg) => arg.trim()).join(' ');
```

Comment:
What if I execute e.g. `node run test "hello world"`? My shell will send `hello world` as a single argument to node (`args` here will consist of a single element), however, the resulting command will be `<script> hello world`, which suddenly has two separate arguments.

Reply:
Shell escaping is tricky to get right and it is different between Posix and Windows.  [Here](https://github.com/npm/promise-spawn/blob/16b36410f9b721dbe190141136432a418869734f/test/escape.js) are the tests for npm's escaping to give you an idea.

Reply:
@anonrig I think you should add the shell escaping tests linked in https://github.com/nodejs/node/pull/52190/files#r1539354437 and https://github.com/npm/promise-spawn/blob/16b36410f9b721dbe190141136432a418869734f/test/escape.js. Shell escaping is. a critical feature of `npm run`

Reply:
Adapted, it's currently 6.5x faster than `npm run test`

```
❯ hyperfine './out/Release/node run test' 'npm run test' -i
Benchmark 1: ./out/Release/node run test
  Time (mean ± σ):      28.4 ms ±   0.6 ms    [User: 22.9 ms, System: 2.9 ms]
  Range (min … max):    27.3 ms …  30.4 ms    95 runs

  Warning: Ignoring non-zero exit code.

Benchmark 2: npm run test
  Time (mean ± σ):     184.7 ms ±  19.0 ms    [User: 136.3 ms, System: 30.7 ms]
  Range (min … max):   175.4 ms … 244.0 ms    12 runs

  Warning: Ignoring non-zero exit code.
  Warning: The first benchmarking run for this command was significantly slower than the rest (244.0 ms). This could be caused by (filesystem) caches that were not filled until after the first run. You should consider using the '--warmup' option to fill those caches before the actual benchmark. Alternatively, use the '--prepare' option to clear the caches before each timing run.

Summary
  ./out/Release/node run test ran
    6.50 ± 0.68 times faster than npm run test
```

Reply:
I don't understand how your latest comment relates to shell escaping.